[FLINK-28915] Flink Native k8s mode jar localtion support s3 schema.

SwimSweet opened the pull request:
## What is the purpose of the change
Flink On Kerbernetes Native K8s Application Mode supports fetching jar from DFS schema(S3, OSS, HDFS, etc.).


## Brief change log
  - Fetch jar from DFS(S3, OSS, HDFS, etc.) before starting flink cluster.

## Verifying this change
This change added tests and can be verified as follows:
  - Remove `testDeployApplicationClusterWithNonLocalSchema` test
  - Added tests that fetch jar from http schema
  - Added tests that fetch jar from file schema
  -  Manually verify the local, file, oss, HDFS with Kerberos, S3 resource.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive):  no
  - Anything that affects deployment or recovery: Native Kubernetes Application Mode:  yes
  - The S3 file system connector: no 

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

SwimSweet commented:
Please help take a look when you are free @wangyang0918 @Aitozi . thx.

Aitozi commented on docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md:
```diff
@@ -97,14 +97,34 @@ COPY /path/of/my-flink-job.jar $FLINK_HOME/usrlib/my-flink-job.jar
+The jar package will be downloaded from filesystem to
+[kubernetes.user.artifacts.base.dir]({{< ref "docs/deployment/config" >}}#kubernetes-user-artifacts-base-dir)/[kubernetes.namespace]({{< ref "docs/deployment/config" >}}#kubernetes-namespace)/[kubernetes.cluster-id]({{< ref "docs/deployment/config" >}}#kubernetes-cluster-id) path in image.
+{{< /hint >}}
+<span class="label label-info">Note</span> `local` schema is also supported .
```
if we use the non-local URI, we do not have to set the init container in pod template (as shown in [link](https://github.com/apache/flink/blob/dca656b922b6c6d350301d8b8f3e60f87b03c381/docs/content.zh/docs/deployment/resource-providers/native_kubernetes.md#example-of-pod-template)) Or we can add comments here to clarify that: if a local URI provided, the jar must be provided in the image or download by a init container.

Aitozi replied:
Besides, I think the `file` schema is not supported (because it's in the client side)

SwimSweet replied:
According to the flink document [https://nightlies.apache.org/flink/flink-docs-release-1.15/zh/docs/deployment/filesystems/overview/](url). `file` schema refers to a file in the local file system.I think `flle` schema  should be supported.

Aitozi replied:
But, I think the `local` points to the file in the container, the `file` points to the jar in the client side. If it's in the client side, we can not get it from the jobmanager container, Right ?

SwimSweet replied:
`file` schema is pointing to the jar in the container. In this case , `local` schema is equal to `file` schema. The difference between `file` and `local` is that `local` will return the jar path directly. We will fetch the jar  from  the path `file`  points to 
to the path that  `kubernetes-user-artifacts-base-dir`/`kubernetes-namespace`/`kubernetes-cluster-id`.
For this ,I have tested.
![fd74ef22a76b5f894b2072484ad9eff](https://user-images.githubusercontent.com/21351694/186449105-61319671-2d6c-478f-a49d-7b68e952bf21.png)
![00e6f8cb62992bf0bbf413feece51de](https://user-images.githubusercontent.com/21351694/186449122-f7a9dddf-d1de-4f49-ae30-1720e174b027.png)
![25939ce0cc994d156d93953e34e7b4c](https://user-images.githubusercontent.com/21351694/186449136-42c88a23-a7e8-4875-b179-963b65790453.png)

Aitozi replied:
From the user side, I think the `local` is different from `file`. My point is that, although we support the filesystem URL, but we still not support user specify a jar on the client side. 

For example, If I want to submit `job.jar` which is at client machine's `/tmp/` dir, It's not supported. I guess this is why we have `local` and `file` two different scheme here. Also cc @wangyang0918 

```
./bin/flink run-application \
    --target kubernetes-application \
    -Dkubernetes.cluster-id=my-first-application-cluster \
    -Dkubernetes.container.image=custom-image-name \
    /tmp/job.jar  
```

SwimSweet replied:
Exactly as you said ,`/opt/job.jar` is equal `file:///opt/job.jar`.This makes it difficult for users to distinguish whether the Jar package is on the client machine or the container.
BTW, are there plans to support submitting user jars from the client? If `kubernetes.rest-service.exposed.type` is set to ClusterIP, it is difficult client connect the jobmanager deployed in k8s.

Aitozi commented on docs/layouts/shortcodes/generated/kubernetes_config_configuration.html:
```diff
@@ -260,5 +260,17 @@
+            <td>The base dir to put the application job artifacts.</td>
+        </tr>
+        <tr>
+            <td><h5>kubernetes.operator.user.artifacts.http.header</h5></td>
```
The option is not suitable with `operator` word here.

Aitozi commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/artifact/HttpArtifactFetcher.java:
```diff
@@ -0,0 +1,71 @@
+        URL url = new URL(uri);
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+
+        // merged session job level header and cluster level header, session job level header take
```
remove this

Aitozi commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesApplicationClusterEntrypoint.java:
```diff
@@ -65,9 +71,13 @@ public static void main(final String[] args) {
         final Configuration configuration =
                 KubernetesEntrypointUtils.loadConfiguration(dynamicParameters);
 
+        PluginManager pluginManager = PluginUtils.createPluginManagerFromRootFolder(configuration);
```
I think it's better to put these also in the try catch block

Aitozi commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -498,6 +498,20 @@ public class KubernetesConfigOptions {
+    public static final ConfigOption<String> KUBERNETES_USER_ARTIFACTS_BASE_DIR =
+            ConfigOptions.key("kubernetes.user.artifacts.base.dir")
+                    .stringType()
+                    .defaultValue("/opt/flink/artifacts")
```
I have a sense that we could also mount an empty dir for this path. By this, if the container restart, the jar resource will kept. It can be optimized later.

SwimSweet replied:
Good Idea. It is a good way to keep the jar resource  without declaring pv and pvc .But  ,I think We need to  add a configuration to control whether `empty dir` is enabled.As far as I know, `empty dir` is forbidden in some user k8s cluster.I will create a ticket and attach a PR for this feature after finishing PR `FLINK-28915`.

Aitozi reviewed (commented):
Thanks for this PR, left some comments. Besides, I think the target branch of this should be the master.

Aitozi commented:
Ps: the CI failed with doc build

SwimSweet commented:
@Aitozi  thanks for your review, I have addressed your comments. The feature that creating `emptyDir`  for `KUBERNETES_USER_ARTIFACTS_BASE_DIR ` has been implemented  in latest commit. I add some test case for it
Please take a look again.

SwimSweet commented:
@flinkbot run azure

SwimSweet commented:
sorry.I have to rebase the  release-1.15 branch to solve the ci problem[1]    .That will add work to cr .

[1] https://issues.apache.org/jira/browse/FLINK-29161

Aitozi commented:
> sorry.I have to rebase the release-1.15 branch to solve the ci problem[1] .That will add work to cr .
> 
> [1] https://issues.apache.org/jira/browse/FLINK-29161

I think You should directly work based the master branch

SwimSweet commented:
> > sorry.I have to rebase the release-1.15 branch to solve the ci problem[1] .That will add work to cr .
> > [1] https://issues.apache.org/jira/browse/FLINK-29161
> 
> I think You should directly work based the master branch

How to migrate the currently modified content to the mater branch? Should I resubmit PR based on master branch? thx

mbalassi commented:
Closing this as the relevant work has been merged as part of #24065.
